subsys/power/power.c

```diff
@@ -152,6 +158,12 @@ enum power_states _sys_suspend(s32_t ticks)
 
 	if (!post_ops_done) {
 		post_ops_done = 1;
+
+#ifdef CONFIG_SYS_PM_DIRECT_FORCE_MODE
```

Comment:
What is the motivation behind having a different behavior here for direct force mode? If there is a need to clear the forced state, maybe it would apply for the standard mode as well. I think it is simpler to keep the behavior consistent across the two modes, unless there is a good reason behind it.

Reply:
@vanti the original documentation says this function will impact all subsequent
 suspend operations until forcing the SYS_POWER_STATE_AUTO state restores normal operation. I'm not clear what's the usage for this operation but I keep it as original design.
```
/**
 * @brief Force usage of given power state.
 *
 * This function overrides decision made by PM policy
 * forcing usage of given power state in all subsequent
 * suspend operations. Forcing the SYS_POWER_STATE_AUTO
 * state restores normal operation.
 *
 * When enabled SYS_PM_DIRECT_FORCE_MODE, this function
 * can only run in thread context. And before the end of
 * the function forced_pm_state is cleared, so it doesn't
 * impact the subsequent suspend operations in this mode.
 *
 * @param state Power state which should be used in all
 *              subsequent suspend operations or the ongoing
 *              suspend operation in direct force mode or
 *              SYS_POWER_STATE_AUTO.
 */
void sys_pm_force_power_state(enum power_states state);
```

Reply:
If we choose to keep the behavior backward-compatible, I think we should stick to it in direct force mode as well (i.e. remove this code block). Right now it looks error-prone - if I started with code that calls sys_pm_force_power_state() using the standard mode, and decide to throw CONFIG_SYS_PM_DIRECT_FORCE_MODE to reduce the latency (which I think is the purpose of the option), all of a sudden the forced state is automatically cleared. This is rather unexpected.

Reply:
@vanti if there is no special use case to force power state in all subsequent suspend operations, I will  change this API to impact only one-shot.

Reply:
If that is the behavior you want, it'd be good to submit it as a separate pull request, and introduce the behavior for all cases. Another possibility is to have two different API: sys_pm_force_power_state() and sys_pm_force_power_state_one_shot() to differentiate the two if you wish to support both behavior.

Reply:
@vanti ok, let's make the decision here and now that sys_pm_force_power_state will only work for one-shot.

Reply:
I am ok with the change in behavior, though would have made the change into at least a separate commit, given it is a different change from direct force mode. Let's see if others have comments regarding it.

Reply:
@vanti I add one more commit to separate them. Thanks